src/DistributedComputations/distributed_transpose.jl

```diff
@@ -0,0 +1,193 @@
+    pack_buffer!    = from == :y ? Symbol(:pack_buffer_, from, to, :(!)) : Symbol(:pack_buffer_, from, :(!))
+    unpack_buffer!  = to == :y ? Symbol(:unpack_buffer_, to, from, :(!)) : Symbol(:unpack_buffer_, to, :(!))
+    
+    buffer = Symbol(buff, :buff)
+    fromfield = Symbol(from, :field)
+    tofield = Symbol(to, :field)
+
+    transpose_name = string(transpose!)
+    to_name = string(to)
+    from_name = string(from)
+
+    pack_buffer_name = string(pack_buffer!)
+    unpack_buffer_name = string(unpack_buffer!)
+
+    @eval begin
+        """
+            $($transpose_name)(pf::TransposableField)
+
+        Transpose the fields in `TransposableField` from a $($from_name)-local configuration
+        (located in `pf.$($from_name)field`) to a $($to_name)-local configuration located
+        in `pf.$($to_name)field`.
+
+        Transpose Algorithm:
+        ====================
+
+        The transpose algorithm works in the following manner
+
+        1. We `pack` the three-dimensional data into a one-dimensional buffer to be sent to the other cores
+           We need to synchronize the GPU afterwards before any communication can take place. The packing is
+           done in the `$($pack_buffer_name)` function.
+
+        2. The one-dimensional is communicated to all the cores using an inplace `Alltoallv!` MPI
+           routine. From the [MPI.jl documentation]():
+
+           Every process divides the Buffer into `Comm_size(comm)` chunks of equal size,
+           sending the j-th chunk to the process of rank j-1. Every process stores the data received from rank j-1 process
+           in the j-th chunk of the buffer.
+
+           ```
+           rank    send buf                             recv buf
+           ----    --------                             --------
+           0      a, b, c, d, e, f       Alltoall      a, b, A, B, α, β
+           1      A, B, C, D, E, F  ---------------->  c, d, C, D, γ, ψ
+           2      α, β, γ, ψ, η, ν                     e, f, E, F, η, ν
+           ```
+
+           The `Alltoallv` function allows chunks of different sizes to be sent to different cores by passing a `count`,
+           for the moment, chunks of the same size are passed, requiring that the ranks divide the number of grid
+           cells evenly.
+
+        3. Once the chunks have been communicated, we `unpack` the received one-dimensional buffer into the three-dimensional
+           field making sure the configuration of the data fits the reshaping. The unpacking is
+           done via the `$($unpack_buffer_name)` function.
+
+        Limitations:
+        ============
+
+        - The tranpose is configured to work only in the following four directions:
+         
+          1. z-local to y-local
+          2. y-local to x-local
+          3. x-local to y-local
+          4. y-local to z-local
+
+          i.e., there is no direct transpose connecting a x-local to a z-local configuration.
```

Comment:
I'm just curious: is this due to some limitation of the packages we're using? Or does it simply turn out that it's easier to do it without an `x`-to-`z` transpose?

Reply:
nono everything is custom so there should be no limitation. This is just a design decision. If we want to allow for direct x -> z transpose we need to store two additional configuration partitioned as

- (x / Ry, y / Rx, z) 
- (x, y / Ry, z / Rx)

what we have at the moment is only 

- (x / Rx, y / Ry, z)
- (x / Rx, y, z / Ry)
- (x, y / Rx, z / Ry)

For the moment, the configuration we have chosen is superior because we do not allow partitioning in z so the flow of the FFT does not require a direct x to z communication (we always have to do the y-FFT and we always start from a z-local configuration).

It starts to make sense thinking about more freedom for stretched direction (in particular when z is stretched) where the algorithm logic follows:
```julia
    transpose_z_to_y!(storage) # copy data from storage.zfield to storage.yfield
    solver.plan.forward.y!(parent(storage.yfield), buffer.y) 
    transpose_y_to_x!(storage) # copy data from storage.yfield to storage.xfield
    solver.plan.forward.x!(parent(storage.xfield), buffer.x)
    transpose_x_to_y!(storage) # copy data from storage.xfield to storage.yfield
    transpose_y_to_z!(storage) # copy data from storage.yfield to storage.zfield
  
    # solver tridiagonal problem then go back...
```
Here having a x -> z configuration would save one transpose, but note that the x configuration that would be required would be a different one than the one we have now because we cannot land in that configuration from a y-local one. So it requires a bit of thinking. It is not relevant for this 3D-FFT solver though, so it is better to add this flexibility (if indeed it is possible) later when needed

Reply:
Got it! Thanks for the explanation. I'll mark this as resolved.